File: Code Matlab/03. Compression/03.1 Matrix Approximation with SVD/EffectOfAlignmentOnImageCompWithSVD.m
clear, clc, close all

M = zeros(201);
M0 = M;
M0(90:110,:) = 1;

figure('Name','Shifted Matrices')
subplot(2,6,1)
hold on
axis off
axis square
imagesc(M0),colormap(gray)

[U,S,V] = svd(M0,'econ');

svalues = diag(S);
subplot(2,6,7)
plot(svalues(1:30),'k-o','LineWidth',2)
    
x = 1:201;
for i = 1:5
    f = round(i*10*sin(2*pi/201*x));
    Mi = M;
    for ii = 1:201
        Mi(90+f(ii):110+f(ii),ii) = i;
    end
    [U,S,V] = svd(Mi,'econ');
    
    subplot(2,6,i+1)
    hold on
    axis off
    axis square
    imagesc(Mi)
    
    svalues = diag(S);
    
    subplot(2,6,i+7)
    plot(svalues(1:30),'k-o','LineWidth',2)
end

alimp = imread('Aligned_Picture.jpg');
unalimp = imread('Unaligned_Picture.jpg');

al = rgb2gray(alimp);
unal = rgb2gray(unalimp);

al = al(1:626,1:500);
unal = unal(200:826,150:650);

figure('Name','Alignment practical')
subplot(2,2,1)
imshow(al)

subplot(2,2,2)
imshow(unal)

al = double(al);
unal = double(unal);

[Ua,Sa,Va] = svd(al,'econ');
[Uu,Su,Vu] = svd(unal,'econ');

svalues_aligned = diag(Sa);
svalues_unaligned = diag(Su);

subplot(2,2,3)
semilogy(svalues_aligned(1:200),'k-o')
hold off

subplot(2,2,4)
semilogy(svalues_unaligned(1:200),'k-o')
hold off

File: Code Python/03. Compression/03.1 Matrix Approximation with SVD/EffectOfAlignmentOnImageCompWithSVD.py
import numpy as np
import matplotlib.pyplot as plt

M = np.zeros((201, 201))
M0 = M
M0[90: 110, :] = 1

fig1 = plt.figure(figsize=(20, 10))
plt.subplot(2, 6, 1)
plt.imshow(M0, cmap='gray')
plt.axis('off')

[U, S, V] = np.linalg.svd(M0, full_matrices=False)
plt.subplot(2, 6, 7)
plt.plot(S[0:29], linewidth=2, color='k', linestyle='-', marker='o')

x = range(1, 202)
for i in range(1, 5):
    f = np.int8(np.round(i * 10 * np.sin(np.dot(2 * np.pi / 201, x))))
    Mi = np.zeros((201, 201))
    for ii in range(1, 201):
        Mi[(90 + f[ii]):(110 + f[ii]), ii] = 1

    plt.subplot(2, 6, i+1)
    plt.imshow(Mi, cmap='gray')
    plt.axis('off')

    [U, S, V] = np.linalg.svd(Mi, full_matrices=False)
    plt.subplot(2, 6, i + 7)
    plt.plot(S[0: 29], linewidth=2, color='k', linestyle='-', marker='o')

plt.show()


